### ISSUE TITLE
Add a correction for oversampling of resistance to adjust map resistance prevalence by drug

### ISSUE THREAD

**Issue by mahafarhat** [farhat-lab/gentb-site#160]
@avdixit please share with @doctormo the global WHO RR/MDR prevalence estimates that you are using to correct our data for oversampling. Please share with him the formula to use the WHO prevalence to correct individual drug frequencies in the map > country> drug interface.

**Comment by avdixit** [comment]
```#R function that takes a country name and generates antibiogram corrected for MDR oversampling. Needs output of WDNN model with drugs as column names and converted into binary based on predetermined thersholds based on Chen et al. 2019,
# list of isolates corresponding to each row of the WDNN model output, geographic metadata with first column labelled 'strain' and second column labelled 'country'
# Also need to provide the prevalence of new MDR/RR for the country from WHO website in the form of probability 

create_antibiogram <- function(country, p_country_mdr, model_out = "model_output_binary.csv", isolates = "isolate_list.txt", geo_metadata = "geo_data_ALL.txt") {
  #read all required files
  model_output <- read.csv(paste(model_out))
  isolate_list <- readLines(paste(isolates))
  rownames(model_output) <- isolate_list
  geog <- read.csv(paste(geo_metadata))
  
  country_isolates <- geog[which(geog$country == country),]
  country_resistant <- model_output[which(rownames(model_output) %in% country_isolates$strain),]
  
  country_resistant_count <- vector()
  
  for (i in 1:ncol(country_resistant)) {
    country_resistant_count[i]  <-  sum(country_resistant[,i])
  }
  
  names(country_resistant_count) <- colnames(country_resistant)
  
  #correct for mdr oversampling
  
  p_a_mdr <- function(a) {
    sum(country_resistant[which(country_resistant$rif == 1),paste(a)])/sum(country_resistant$rif)
  }
  
  p_mdr_a <- function(a) {
    sum(country_resistant[which(country_resistant[,paste(a)] ==1), "rif"])/sum(country_resistant[,paste(a)])
  }
  
  p_a_res <- function(a) {
    (p_a_mdr(a)*p_country_mdr)/p_mdr_a(a)
  }
  
  corrected_count <- vector()
  
  for (i in 1:ncol(country_resistant)) {
    corrected_count[i] <- p_a_res(colnames(country_resistant)[i])
  }
  names(corrected_count) <- colnames(country_resistant)
  antibiogram_corrected <- barplot(corrected_count*100, ylim = c(0.0,max(corrected_count*100)+5), main = paste("Antibiogram for", country, "(corrected)", sep = " "), xlab = "Antibiotic", ylab = "Percent of resistant isolates", col = "orange")
  text(antibiogram_corrected, y = 0, round(corrected_count*100, 1), cex = 1, pos = 3, col = "black")
}
```

**Comment by avdixit** [comment]
[wdnn_new.zip](https://github.com/farhat-lab/gentb-site/files/3325749/wdnn_new.zip)
WDNN Model re-trained using Keras 2.2.4 and tensorflow 1..0

**Comment by avdixit** [comment]
Sample WDNN model output: [model_output_new.txt](https://github.com/farhat-lab/gentb-site/files/3325759/model_output_new.txt)
Isolate names to be used as row names for the model output above [isolate_names.txt](https://github.com/farhat-lab/gentb-site/files/3325760/isolate_names.txt)

**Comment by avdixit** [comment]
```
#This code takes model output and converts it into a binary output (0 for susceptible and 1 for resistant) based on thresholds described in Chen at al. EBioMedicine. 2019.

model_output <- read.csv("model_output_new.txt")
isolate_list <- readLines("isolate_names.txt")
rownames(model_output) <- isolate_list
model_output_binary <- subset(model_output, select = -c(cip)) #remove ciprofloxacin because phenotypic data is missing for cipro

model_output_binary$rif <- as.numeric(model_output$rif < 0.35)
model_output_binary$inh <- as.numeric(model_output$inh < 0.09)
model_output_binary$pza <- as.numeric(model_output$pza < 0.32)
model_output_binary$emb <- as.numeric(model_output$emb < 0.40)
model_output_binary$str <- as.numeric(model_output$str < 0.26)
model_output_binary$cap <- as.numeric(model_output$cap < 0.23)
model_output_binary$amk <- as.numeric(model_output$amk < 0.20)
model_output_binary$moxi <- as.numeric(model_output$moxi < 0.39)
model_output_binary$oflx <- as.numeric(model_output$oflx < 0.57)
model_output_binary$kan <- as.numeric(model_output$kan < 0.32)

write.csv(model_output_binary, "model_output_binary.csv", quote = FALSE, sep = "\t", row.names = FALSE)
```

**Comment by mahafarhat** [comment]
Thanks @avdixit but R-code will not be compatible with genTB. So I've masked the few prior inputs you've added. Please instead provide an algebraic formula to correct/scale the resistance frequency for a particular drug and country, and the link to the latest WHO MDR/RR rates by country.

**Comment by avdixit** [comment]
Here is the algebraic formula 
<img width="828" alt="image" src="https://user-images.githubusercontent.com/20425908/60131356-2e8e1580-9767-11e9-8ce2-a18b79dbe9e7.png">

Link to WHO MDR/RR rates by country: https://www.who.int/tb/country/data/profiles/en/
After selecting the country, use the "Estimated % of TB cases with MDR/RR-TB" for "New cases"

**Comment by mahafarhat** [comment]
Thanks Avika, and we should note that the P(rifampicin= R) = WHO estimate of Rifampicin resistance prevalence. 

Maha

**Comment by doctormo** [comment]
Raw csv data is being sourced from: https://www.who.int/tb/country/data/download/en/

**Comment by doctormo** [comment]
The feature has been added and deployed live, please see https://gentb.hms.harvard.edu/maps/ and select only ONE country (it does not show for all or multiple countries)

**Comment by mahafarhat** [comment]
Thanks @doctormo a couple of points of feedback. The blue color for oversampling is too close to other blue for sensitive intermediate and MDR. Also I'm not exactly clear how you are generating the oversampling plot. As is the plot set up implies that "oversampling" is a separate frequency, similar to sensitive, or MDR etc. but the reality is that we have R-raw (raw resistance frequency), and R-corrected (corrected for oversampling). should we discuss further how to display this more simply?

**Comment by doctormo** [comment]
There are some harsh limitations on how we can display the data. When we talked about it last, we'd roughly brought it to this design, with oversampling as a seperate key that can be toggled using the nv internal mechanisms. (otherwise we'd have to implement our own)

**Comment by mahafarhat** [comment]
Unfortunately it’s very confusing as is and the data will be miss interpreted. Perhaps the toggle can be separated on the left hand side and a starkly different color can be used. Also I’m pretty sure the calculations are not accurate as the y axis is up to 14thousand from about 900 without got Peru. That’s just not possible. 

Maha

**Comment by mahafarhat** [comment]
@doctormo please remove the oversampling correction from live ASAP. it's default at the moment when one selects drug, and the information given is incorrect.

**Comment by mahafarhat** [comment]
@doctormo oversampling correction is back on live (was previously removed) not sure why. Please remove it.

**Comment by doctormo** [comment]
It's not on live, it's only visible if you're an admin. If you log out, you should see it without the numbers.
